src/pages/staining/ComplexStainForm.tsx

```diff
@@ -0,0 +1,255 @@
+import React from "react";
+import {
+  ComplexStainRequest,
+  RecordComplexStainMutation,
+  StainPanel,
+} from "../../types/sdk";
+import { Form, Formik } from "formik";
+import FormikInput from "../../components/forms/Input";
+import GrayBox, { Sidebar } from "../../components/layouts/GrayBox";
+import { motion } from "framer-motion";
+import variants from "../../lib/motionVariants";
+import Heading from "../../components/Heading";
+import MutedText from "../../components/MutedText";
+import LabwareScanPanel from "../../components/labwareScanPanel/LabwareScanPanel";
+import columns from "../../components/dataTable/labwareColumns";
+import PinkButton from "../../components/buttons/PinkButton";
+import * as Yup from "yup";
+import { useMachine } from "@xstate/react";
+import createFormMachine from "../../lib/machines/form/formMachine";
+import { reload, stanCore } from "../../lib/sdk";
+import FormikSelect from "../../components/forms/Select";
+import { objectKeys } from "../../lib/helpers";
+import { FormikLabwareScanner } from "../../components/labwareScanner/FormikLabwareScanner";
+import Table, {
+  TableBody,
+  TableCell,
+  TableHead,
+  TableHeader,
+} from "../../components/Table";
+import WorkNumberSelect from "../../components/WorkNumberSelect";
+import OperationCompleteModal from "../../components/modal/OperationCompleteModal";
+import Warning from "../../components/notifications/Warning";
+
+type ComplexStainFormValues = ComplexStainRequest;
+
+type ComplexStainFormProps = {
+  stainType: string;
+};
+
+export default function ComplexStainForm({ stainType }: ComplexStainFormProps) {
+  const [current, send] = useMachine(
+    createFormMachine<
+      ComplexStainRequest,
+      RecordComplexStainMutation
+    >().withConfig({
+      services: {
+        submitForm: (ctx, e) => {
+          if (e.type !== "SUBMIT_FORM") return Promise.reject();
+          return stanCore.RecordComplexStain({ request: e.values });
+        },
+      },
+    })
+  );
+
+  const { serverError } = current.context;
+
+  const plexMin = 1;
+  const plexMax = 100;
+
+  const validationSchema = Yup.object().shape({
+    stainType: Yup.string().required().label("Stain Type"),
+    plex: Yup.number()
+      .integer()
+      .min(plexMin)
+      .max(plexMax)
+      .required()
+      .label("Plex Number"),
+    panel: Yup.string()
+      .oneOf(Object.values(StainPanel))
+      .required()
+      .label("Experimental Panel"),
+    labware: Yup.array()
+      .of(
+        Yup.object().shape({
+          barcode: Yup.string().required().label("Barcode"),
+          bondBarcode: Yup.string().required().label("Bond Barcode"),
+          bondRun: Yup.number().integer().positive().label("Bond Run"),
```

Comment:
In acceptance test it says -bond run number - 4 digit integer (+ve) . Is it required  to put 4 digit number constraint here?

Reply:
I thought that might be a bit too restrictive for the client. Seemed like the sort of thing that may need to change in the relatively near future, so thought I'd leave that as core side validation only.